### PR TITLE
Add interpreted (recursive) functions

### PR DESCRIPTION
**Pull request by skius**
This PR adds wrapping to the `z3` crate of `z3-sys`'s `Z3_mk_rec_func_decl` and `Z3_add_rec_def`.

I implemented a new `RecFuncDecl` struct which has an `add_def(args, body)` method, since the `z3-sys` API differentiates between `mk_func_decl` and `mk_rec_func_decl` as far as I can tell.

The `add_def` method has a doc example on how to use the new API and there are two tests (`test_rec_func_def` and `test_rec_func_def_unsat`) in `z3/test/lib.rs`.

I'm not familiar at all with unsafe Rust code and only have a very basic understanding of SMT solvers, so please take a good look and let me know if something's not up to standard.

### CODE REVIEW

File: z3/src/rec_func_decl.rs
Comment (sameer) [inline z3/src/rec_func_decl.rs:79]: do we need to assert on args & body being in the same ctx?

File: z3/src/rec_func_decl.rs
Comment (sameer) [inline z3/src/rec_func_decl.rs:178]: This might need a mutex lock -- not sure that z3 reference counters are atomic.

**Review by sameer — CHANGES_REQUESTED** [review]
Hi Niels,

Thanks for the PR, this looks really good. Just 2 minor comments and we can merge this.

Best,
Sameer

File: z3/src/rec_func_decl.rs
Comment (skius) [reply z3/src/rec_func_decl.rs:178]: I'm not sure either, I copied that impl from FuncDecl https://github.com/prove-rs/z3.rs/blob/8b77fc554def507162a3860bea659247f2939859/z3/src/func_decl.rs#L128-L134 - should I aquire Z3_MUTEX?

File: z3/src/rec_func_decl.rs
Comment (skius) [reply z3/src/rec_func_decl.rs:79]: I can't comment on this, I don't really know where/how one would use multiple contexts.

File: z3/src/rec_func_decl.rs
Comment (sameer) [reply z3/src/rec_func_decl.rs:180]: Ah ok, that's fine then. We can continue that thread in #143

**Review by sameer — APPROVED** [review]
LGTM! Will merge after the workflow approves

**Comment by skius** [comment]
Great, thank you!

**Pull request merged**
